### ISSUE TITLE
Enum max values can't be used to dimension an array

### ISSUE THREAD

**Issue by gpersoon** [ethereum/solidity#12886]
## Description

Enum max values can't be used to dimension an array.
An enum max value can be assigned to a constant. But that constant can't be used as an array size.
'Normal' constants can be used to dimension an array.
See example below.

## Environment
- Compiler version: 0.8.13
- Target EVM version (as per compiler settings): default
- Framework/IDE (e.g. Truffle or Remix): Remix 
- EVM execution environment / backend / blockchain client:
- Operating system: Windows

## Steps to Reproduce

```solidity
//SPDX-License-Identifier: MIT
pragma solidity 0.8.13;

contract test {
    enum ParameterIndexes { P1,P2 }
    uint constant x1 = uint(type(ParameterIndexes).max);
    uint256[x1] params1; // Doesn't compile: Error expected integer literal or constant expression

    uint256[uint(type(ParameterIndexes).max)] params2; // Doesn't compile: Error expected integer literal or constant expression

    uint constant x2 = 7;
    uint256[x2] params3; // Does compile
}
```

**Comment by StrongerXi** [comment]
Please see #12832

**Comment by StrongerXi** [comment]
If there is a strong use case for this, maybe we should consider making `ConstantEvaluator` more aggressive to handle these cases? i.e., elementary type casting on constant expression, and expression like `type(...).max`.

**Comment by hrkrshnn** [comment]
To explain what's going on here, array length should be known during the beginning of the analysis time. However, the expressions `type(Enum).max` and `uint(type(Enum).max)` are currently only known towards the end of the code generation. To do this properly, we need better support for compile time constant evaluation. See https://github.com/ethereum/solidity/issues/3157. However, maybe it is possible to create special cases for `uint(...)` and `type(Enum).max` to deal with this easily. Similar examples are `type(uint8).max`.

**Comment by cameel** [comment]
I don't think this really needs #3157 to be done properly. It's related but the problem seems to be not in evaluating it (you can already assign `type(...).max` to a constant and it gets evaluated at compile time) but in the analysis being aware that it's a constant at the point where it's used to size an array.

I think it could be solved in a way similar to #12928, i.e. by just marking the expression as pure. May also need moving the evaluation of `type(...).max` earlier in the analysis but this seems like something that we should be able to evaluate as soon as we have the enum definition parsed.
